src/server/system_services/master_key_manager.js

```diff
@@ -0,0 +1,246 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const _ = require('lodash');
+const util = require('util');
+const crypto = require('crypto');
+const config = require('../../../config');
+const db_client = require('../../util/db_client').instance();
+const dbg = require('../../util/debug_module')(__filename);
+const SensitiveString = require('../../util/sensitive_string');
+const LRUCache = require('../../util/lru_cache');
+
+const ROOT_KEY = 'noobaa-root-master-key';
+
+class MasterKeysManager {
+    constructor() {
+        this.is_initialized = false;
+        this.resolved_master_keys_by_id = {};
+        this.master_keys_by_id = {};
+        this.secret_keys_cache = new LRUCache({
+            name: 'SecretKeysCache',
+            expiry_ms: Infinity,
+            max_usage: 5000,
+            make_key: params => params.encrypted_value,
+            load: params => new SensitiveString(params.decipher.update(
+                Buffer.from(params.encrypted_value, 'base64')).toString())
+        });
+    }
+
+    static get_instance() {
+        MasterKeysManager._instance = MasterKeysManager._instance || new MasterKeysManager();
+        return MasterKeysManager._instance;
+    }
+
+    /**
+     * 
+     * @returns {String} 
+     */
+    get_root_key_id() {
+        return ROOT_KEY;
+    }
+
+    /**
+     * 
+     * @returns {Object} 
+     */
+    get_root_key() {
+        return this.resolved_master_keys_by_id[ROOT_KEY];
+    }
+
+    async load_root_key() {
+        if (this.is_initialized) return;
+        const { NOOBAA_ROOT_SECRET } = process.env;
+        if (!NOOBAA_ROOT_SECRET) throw new Error('NON_EXISTING_ROOT_KEY');
+        const r_key = {
+            _id: 'noobaa-root-master-key',
+            cipher_key: Buffer.from(NOOBAA_ROOT_SECRET, 'base64'),
+            cipher_type: "aes-256-gcm",
+            description: "Root Master Key",
+            disabled: false
+        };
+        this.resolved_master_keys_by_id[ROOT_KEY] = r_key;
+        this.is_initialized = true;
+    }
+
+    /**
+     * 
+     * @param {Array} master_keys_by_id 
+     */
+    update_master_keys(master_keys_by_id) {
+        if (!master_keys_by_id) {
+            dbg.error('Invalid master_keys_by_id, got ', master_keys_by_id);
+            throw new Error('Invalid master_keys_by_id');
+        }
+        this.master_keys_by_id = master_keys_by_id;
+    }
+
+    /**
+     * 
+     * @param {Object} options 
+     * @returns {Object} 
+     */
+    new_master_key(options) {
+        const { description, master_key_id, cipher_type } = options;
+        const _id = db_client.new_object_id();
+        const m_key = _.omitBy({
+            _id,
+            description,
+            master_key_id: master_key_id && master_key_id !== ROOT_KEY ? db_client.parse_object_id(master_key_id) : undefined,
```

Comment:
This causes the system to not have master_key_id which is weird since it's master key is the root key.